### PR TITLE
Fix the order in which the wrapper parses properties

### LINKED ISSUE #985: GradleWrapperMain loads system properties in wrong order
`GradleWrapperMain` loads system properties in the reverse order than the docs suggest and as the "main" gradle process does.

### Expected Behavior
ill quote the docs:
> The configuration is applied in following order (if an option is configured in multiple locations the last one wins):
>    from gradle.properties in project build dir.
>    from gradle.properties in gradle user home.
>    from system properties, e.g. when -Dsome.property is set on the command line.

### Current Behavior
during the `gradlew` bootstrapping process, system properties (`systemProp.*`) defined in `gradle.properties` in the project dir override those defined in `gradle.properties` in the user home directory:

https://github.com/gradle/gradle/blob/v3.2.1/subprojects/wrapper/src/main/java/org/gradle/wrapper/GradleWrapperMain.java#L69

and those defined in the user home `gradle.properties` override command-line supplied system props (`-Dxyz`):

https://github.com/gradle/gradle/blob/v3.2.1/subprojects/wrapper/src/main/java/org/gradle/wrapper/GradleWrapperMain.java#L52

### Context
use-case:
* default http proxy defined in `<projectDir>/gradle.properties`
* you are using an override of the http proxy defined in `<userHome>/.gradle/gradle.properties`
* `wrapper.gradleVersion` is updated, wrapper is regenerated
* gradlew launch wants to download the updated gradle version
* timeout / unknown host

### Steps to Reproduce (for bugs)
n/a

### Your Environment
gradle-3.2.1 (but looking at the commit logs, this behavior seems to be existing for a _very_ long time?)

### PR DESCRIPTION
**Pull request by sschuberth**
Fixes #985 and #11864.

### Context
There were already two issues reported about this, and the fix was easy, so this should be done esp. now as the stale bot closed / is about to close the referenced issues.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy [hunk 1]
```diff
@@ -72,7 +72,7 @@ class WrapperGenerationIntegrationTest extends AbstractIntegrationSpec {
 
         then: "the checksum should be constant (unless there are code changes)"
         // TODO If you need to update this because of necessary code change in the wrapper, also address the TODO in 'wrapper.gradle.kts'
-        sha256(file("first/gradle/wrapper/gradle-wrapper.jar")).asHexString() == "e996d452d2645e70c01c11143ca2d3742734a28da2bf61f25c82bdc288c9e637"
+        sha256(file("first/gradle/wrapper/gradle-wrapper.jar")).asHexString() == "810cd5fb51fa916cd09c9332f5f60bdfd786871adb90e29a2b65cd74cf9730eb"
```
Comment (sschuberth) [inline subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy:75]: BTW, I did not get what the `also address the TODO in 'wrapper.gradle.kts'` comment refers to, as I did not find a `wrapper.gradle.kts` file in the code base. I only found `gradlebuild.wrapper.gradle.kts`, but that file does not contain a `TODO`.

File: subprojects/wrapper/src/main/java/org/gradle/wrapper/GradleWrapperMain.java [hunk 2]
```diff
@@ -66,9 +66,10 @@ public static void main(String[] args) throws Exception {
                 new BootstrapMainStarter());
     }
 
-    private static void addSystemProperties(File gradleHome, File rootDir) {
-        System.getProperties().putAll(SystemPropertiesHandler.getSystemProperties(new File(gradleHome, "gradle.properties")));
```
Comment (oehme) [inline subprojects/wrapper/src/main/java/org/gradle/wrapper/GradleWrapperMain.java:70]: I found the code clearer before, i.e. explicitly calling System.getProperties to make it obvious what we're changing here.

File: subprojects/wrapper/src/main/java/org/gradle/wrapper/GradleWrapperMain.java [hunk 3]
```diff
@@ -66,9 +66,10 @@ public static void main(String[] args) throws Exception {
...
-        System.getProperties().putAll(SystemPropertiesHandler.getSystemProperties(new File(rootDir, "gradle.properties")));
+    private static void addSystemProperties(Properties systemProperties, File gradleHome, File rootDir) {
```
Comment (oehme) [inline subprojects/wrapper/src/main/java/org/gradle/wrapper/GradleWrapperMain.java:69]: Should be called `gradleUserHome`

File: subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc [hunk 4]
```diff
@@ -32,15 +32,16 @@ Aside from configuring the build environment, you can configure a given project
...
+Gradle provides several options that make it easy to configure the Java process that will be used to execute your build. While it's possible to configure these in your local environment via `GRADLE_OPTS` or `JAVA_OPTS`, it is useful to be able to store certain settings like JVM memory configuration and Java home location in version control so that an entire team can work with a consistent environment. To do so, place these settings into a `gradle.properties` file committed to your version control system.
 
-Setting up a consistent environment for your build is as simple as placing these settings into a `gradle.properties` file. The configuration is a combination of all your `gradle.properties` files, but if an option is configured in multiple locations, the _first one wins_:
+The final configuration taken into account by Gradle is a combination of all your `gradle.properties` files. If an option is configured in multiple locations, the _first one_ found in any of these locations wins:
 
-* system properties, e.g. when `-Dgradle.user.home` is set on the command line.
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc:39]: Shouldn't be removed, system properties on the command line always win. This is not just about `-Dgradle.user.home` (which is a bad example), but any system prop you wanna pass to the build.

**Review by oehme — CHANGES_REQUESTED** [review]
Just a remark on the documentation and some nit-picks on the code. Looks good overall.

File: subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc (hunk 4 above)
Comment (sschuberth) [reply subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc:39]: To me, it was confusing why system properties are mentioned in this section to start with. This section is about *Gradle* properties after all. I know that you can also set system properties inside `gradle.properties`, so in a way, system properties can be seen as a subset of Gradle properties. And it does not make sense to me to limit the docs to a subset of Gradle properties here.

File: subprojects/wrapper/src/main/java/org/gradle/wrapper/GradleWrapperMain.java (hunk 3 above)
Comment (sschuberth) [reply subprojects/wrapper/src/main/java/org/gradle/wrapper/GradleWrapperMain.java:69]: Probably. Maybe. But that's out of scope of this change, as the variable was called like that before.

File: subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc (hunk 4 above)
Comment (oehme) [reply subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc:39]: You can mention `-P` then. I just want it to be clear that the command line wins over all the files.

File: subprojects/wrapper/src/main/java/org/gradle/wrapper/GradleWrapperMain.java (hunk 2 above)
Comment (sschuberth) [reply subprojects/wrapper/src/main/java/org/gradle/wrapper/GradleWrapperMain.java:70]: To me, "systemProperties" makes just as clear as "System.getProperties()" that we are dealing with system properties here.

File: subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc (hunk 4 above)
Comment (sschuberth) [reply subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc:39]: Done.

File: subprojects/wrapper/src/main/java/org/gradle/wrapper/GradleWrapperMain.java (hunk 3 above)
Comment (sschuberth) [reply subprojects/wrapper/src/main/java/org/gradle/wrapper/GradleWrapperMain.java]: I've anyway added a new commit doing the rename.

File: subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy (hunk 1 above)
Comment (oehme) [reply subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy:75]: Yeah, I don't know where this is supposed to be either. @jjohannes?

**Review by oehme — APPROVED** [review]
@jjohannes this looks good

File: subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy (hunk 1 above)
Comment (jjohannes) [reply subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy:75]: Sorry the file was renamed. This is about `subprojects/wrapper/build.gradle.kts`.

https://github.com/gradle/gradle/blob/524b568298db23f84216840fa10a6f0177ce5248/subprojects/wrapper/build.gradle.kts#L50-L64

**Review by jjohannes — COMMENTED** [review]
I have no opinion on this, but we need to be aware that any change in the wrapper Jar will cause thousands of wrapper Jars pushed to Git repos to be upgrade.

But if this is a critical enough bugfix, I assume that it is ok.

**Comment by oehme** [comment]
Well, it's wrong compared to how Gradle normally reads those properties, so I think that's worth fixing.

**Review by jjohannes — CHANGES_REQUESTED** [review]
True. Let us see if we can get this in.

@sschuberth thanks for pushing this and providing the PR. Do you think you might be able to address this: https://github.com/gradle/gradle/pull/16519/files#r611567851 ?

File: subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy (hunk 1 above)
Comment (sschuberth) [reply subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy:75]: I'm not quite sure I understand what the goal here is. To filter out "gradle-cli-classpath.properties" and "gradle-cli-parameter-names.properties" in the `executableJar` task so they do not get packaged to "gradle-wrapper.jar"? If so, what about the also packaged "gradle-wrapper-classpath.properties" and "gradle-wrapper-parameter-names.properties" files, should they stay?

Comment (jjohannes) [reply subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy:75]: You can just remove the whole "TODO" block I linked above. The goal of it was to not change the Jar. These "properties" are currently in the Jar by mistake (caused by a broken setup that is now fixed). They are not needed at all.

Now, as the Jar changes anyway, we can remove the whole workaround. The only thing you need to address in code is the compiler warning you get in `org.gradle.wrapper.Install` after removing the block.

Comment (sschuberth) [reply subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy:74]: Ok, done.

Comment (jjohannes) [reply subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy:74]: Thanks @sschuberth! Can you also adjusts the `sha256` (see the line this comment thread is attached to) in the test again as it has now changed again due to the additional changes. Right now this test is failing.

Comment (jjohannes) [reply subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy:74]: You will het the right sha by running the test and looking at the test failure.

Comment (sschuberth) [reply subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy:74]: Yes, sure will do that later. Sorry for forgetting that.

Comment (sschuberth) [reply subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy]: Done.

**Comment by jjohannes** [comment]
@bot-gradle test this

File: subprojects/wrapper/src/main/java/org/gradle/wrapper/GradleWrapperMain.java (hunk 2 above)
Comment (jjohannes) [reply subprojects/wrapper/src/main/java/org/gradle/wrapper/GradleWrapperMain.java:70]: Thanks! Sorry for the late reply (I was out for a bit longer). I'll see that we can get this merged now.

**Review by jjohannes — APPROVED** [review]

**Comment by jjohannes** [comment]
@bot-gradle test this

**Comment by jjohannes** [comment]
@bot-gradle test and merge

**Comment by jjohannes** [comment]
@bot-gradle test and merge

**Comment by sschuberth** [comment]
Are you fighting flaky tests, or is there still a problem with my change @jjohannes?

**Comment by jjohannes** [comment]
@sschuberth yes there was some flakiness first, but there also seems to be an issue a link in the docs. If you don't mind, can you:

- Rebase to latest `master`
- Run `:docs:checkDeadInternalLinks` and fix the issue (if there is any). CI seems to find one, but I can't see what it is right now.

**Comment by sschuberth** [comment]
>     * Rebase to latest `master`

Done.

>     * Run `:docs:checkDeadInternalLinks` and fix the issue (if there is any). CI seems to find one, but I can't see what it is right now.

Also done, there indeed was a dead link (which I fixed by actually adding a marker to the section I wanted to link to).

**Comment by jjohannes** [comment]
@sschuberth ah yes. In favour of having the check and not breaking existing links when changing section titles, we decided to have markers for every section, although in practice the linking could work without. 

Thanks for fixing this ans sorry for the trouble.

**Comment by jjohannes** [comment]
@bot-gradle test and merge

**Pull request merged**